lib/puppet/parser/compiler/catalog_validator/relationship_validator.rb

```diff
@@ -30,7 +30,7 @@ def validate_relationship(param)
         # all other relationships requires the referenced resource to exist when mode is strict
         refs = param.value.is_a?(Array) ? param.value.flatten : [param.value]
         refs.each do |r|
-          next if r.nil?
+          next if r.nil? || r == :undef
```

Comment:
The 3.x (wonky, bat-shit-crazy) API is different depending on where an `undef` is - sometimes it is nil, sometimes it is `:undef`. You can observe the difference by using `a => undef` vs. `a => [undef]`. (This also answers the question above, why value is not :undef when it is not in an array).